Add CCA feature files for DIP creation and upload

jraddaoui opened the pull request:
Includes two feature files to describe the requirements of two new scripts
in the automation-tools project, one to create a DIP from an AIP in the SS
and the other to upload a DIP folder to an AtoM instance. Only the
create_CCA_dip.feature file is fully automated.

See https://github.com/artefactual-labs/archivematica-acceptance-tests/issues/29

jrwdunham commented:
I think this needs a rebase now, after https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/44

jraddaoui commented:
Done!

sevein commented on archivematicaselenium.py:
```diff
@@ -1271,7 +1271,7 @@ def _parse_tasks_table_am_1_7(self, tasks_url, table_dict):
                 next_tasks_url = '{}{}'.format(
                     self.am_url, link_button.get_attribute('href'))
-        self.driver.close()
+        self.driver.quit()
```
Please remember to check with Joel if this is a good idea.

jrwdunham replied:
I'm fine with the change to `quit()` now. Besides, it's already been done in a recently merged PR. Speaking of which, this needs to be rebased before it can be merged.

sevein commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -0,0 +1,40 @@
+@am16 @create_CCA_dip @automation-tools @wip
```
I think we need some input from @jrwdunham on which tags should be used.

jraddaoui commented:
I have rebased and removed the WIP tag.

jraddaoui commented:
Hi @jrwdunham, did you have a chance to take a look at this?

jrwdunham commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -0,0 +1,40 @@
+  #   "./image.png",1510647651
+  #   "./folder1/file 1.txt",1510647659
+  #   "./folder2/file 2.txt",
+  #
```
I see what you're doing here, but my criticism is that this is not really a fully automated test if you need to manually get Archivematica into a specific state beforehand, right? I mean, there is precedent for `Given`s that create AIPs. You could modify one of these to create your AIP and then you would not need to pass `aip_uuid` and `transfer_name` as userdata and the test would be more modifiable. See `@given('an encrypted AIP in the standard GPG-encrypted space')` in steps.py. (I realize you may not have time to do this, or there may not be a need. Just putting it out there.)

jrwdunham commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -0,0 +1,40 @@
+@am16 @create_CCA_dip @automation-tools
```
Can you please provide comments at the top of this file which detail how you were able to get this feature file to run (and pass)? I think we should be doing this for all new feature files (and adding such comments to existing ones.) Details needed:

- Description of the Archivematica deploy.
- Full behave command used (with idiosyncratic paths obscured), e.g., `behave --tags=create_CCA_dib --no-skipped -D files_csv=/path/to/file.csv`.
- Content of the actual CSV file.

jrwdunham replied:
Nice!

jrwdunham commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -0,0 +1,40 @@
+  Scenario: Creating DIP with automation tools
+    Given Tim has configured the automation-tools DIP creation bash script with all the required parameters
+    And he has created that AIP using the current version of Archivematica (1.6.x)
+    When he initiates the DIP creation script execution
```
I think "initiates" is inaccurate here, no? I mean the `check_output` blocks until the script completes, no? In which case I think the wording of the following `Then`s can be more clear. Right now it's a bit confusing that there is an implicit "the script" in all of these `Then`s, e.g., `"creates a METS XML file that describes the contents of the DIP"` is implicitly `"the script creates a METS XML file that describes the contents of the DIP"`. I think it would be clearer if these were `"the DIP METS XML file describes the contents of the DIP"`, `"the DIP contains an objects directory containing one zip container with all objects from the original transfer"`, etc.

jraddaoui commented on features/steps/steps.py:
```diff
@@ -1544,6 +1547,129 @@ def step_impl(context):
+    context.zip_path = os.path.join(objects_path,
+                                    '{}.zip'.format(context.transfer_name))
+    assert os.path.exists(context.zip_path)
+    context.zip_file = zipfile.ZipFile(context.zip_path, 'r')
```
I don't think it's a big deal, but I like it better than closing the `context.zip_file` in the last step. Thanks!

jraddaoui commented on features/steps/steps.py:
```diff
@@ -1544,6 +1547,129 @@ def step_impl(context):
+    assert files
+    # Check file info in ZIP files
+    for file_info in context.zip_file.infolist():
+        # Strip transfer name and '/', the main folder will end empty
```
The main folder filename in the ZIP file is `transfer_name/` and we're stripping that part from it, so it will end as an empty string.

jraddaoui commented on features/steps/steps.py:
```diff
@@ -1544,6 +1547,129 @@ def step_impl(context):
+            continue
+        lastmodified = int(time.mktime(file_info.date_time + (0, 0, -1)))
+        # Find file by filename in file info from CSV
+        csv_info = next((x for x in files if x['filename'] == filename), None)
```
Thanks! Javascript thinking xD

jraddaoui commented:
Thanks @jrwdunham! Great feedback.

I think I've addressed most of it, except the part about creating an AIP to allow the scenario to pass without requiring one already created in AM/SS. I thought about it and had a conversation with @jhsimpson where we decided not to do it for now. Sorry, I forgot to mention it in the PR. I added a note related to this in https://github.com/artefactual-labs/archivematica-acceptance-tests/issues/49, so I hope we can do something in the future.

I have also:
  - Rebased the master branch
  - Found and fixed an issue with the Chromedriver
  - Included a related feature file
  - Updated the PR subject and description

Please, could you take another look?

jrwdunham commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -24,17 +24,17 @@ Feature: Create CCA Dissemination Information Package
-    And includes a copy of the submissionDocumentation included in the original Transfer
-    And includes a copy of the AIP METS file generated during ingest
-    And stores the DIP locally in the output directory specified in the script parameters
+    And the DIP METS XML file that describes the contents of the DIP
```
Grammar here. I think maybe it should be "And the DIP METS XML file describes the contents of the DIP"?

jrwdunham commented on features/CCA/create_CCA_dip.feature:
```diff
@@ -24,17 +24,17 @@ Feature: Create CCA Dissemination Information Package
+    And the DIP objects directory contains one zip container with all objects from the original transfer
+    And each DIP object file has its original filename and last modified date from the original transfer
+    And the DIP zip file includes a copy of the submissionDocumentation from the original Transfer
+    And a copy of the AIP METS file generated during ingest
```
Are the above two `Then`s  logically one? That is, does the second one actually mean `And [the DIP zip file includes] a copy of the AIP METS file generated during ingest`? If so, I think you could clarify that either by indenting the second `Then` or adding the text in brackets above.

jrwdunham commented on features/core/upload-DIP.feature:
```diff
@@ -0,0 +1,28 @@
+    
+# we mention the script parameters in the Given step and the "when" step - should this 
+# just be done once? I am not sure exactly how the parameters will be provided so 
+# wasn't sure which step it should go in.
```
I think it's fine as is. When/if implemented, these parameters could be placed in a step data table, accessible as `context.table`. See https://pythonhosted.org/behave/tutorial.html:

     And configured the automation-tools DIP upload script with all the required parameters
       | atom_url      | atom_api_key |
       | http://...    | ...          |

jrwdunham commented on features/CCA/upload_CCA_dip.feature:
```diff
@@ -0,0 +1,21 @@
+#
+# The scenario included in this feature has not been automated yet.
+
+@am16 @upload_CCA_dip @automation-tools
```
Why was the `@wip` tag removed? I think we should adopt a convention for tagging features and scenarios that are non-executable, i.e., designed to be documentation only. I suggest `@non-executable`. Thoughts? If we use that, maybe this should be tagged with it.

jraddaoui commented:
Great @jrwdunham!

I saw those tables when I was doing the first feature file, but I thought it was better to use user data as this tables look better to test different cases in the same scenario. The comments from the second file are from @joel-simpson but, as we decided not to automate it, I removed them.

`@non-executable` works for me, I'll do the couple of changes and merge.

Thanks!!
